integration/spark/src/main/scala/org/apache/spark/sql/execution/command/index/ShowIndexesCommand.scala

```diff
@@ -54,69 +62,87 @@ case class ShowIndexesCommand(
     CarbonInternalMetastore.refreshIndexInfo(
       carbonTable.getDatabaseName, tableName, carbonTable)(sparkSession)
-    val indexesMap = CarbonInternalScalaUtil.getIndexesMap(carbonTable)
-    if (null == indexesMap) {
-      throw new Exception("Secondary index information is not loaded in main table")
-    }
-    val indexTableMap = indexesMap.asScala
-    if (indexTableMap.nonEmpty) {
-      val indexList = indexTableMap.map { indexInfo =>
-        try {
-          val isSITableEnabled = sparkSession.sessionState.catalog
-            .getTableMetadata(TableIdentifier(indexInfo._1, dbNameOp)).storage.properties
-            .getOrElse("isSITableEnabled", "true").equalsIgnoreCase("true")
-          if (isSITableEnabled) {
-            (indexInfo._1, indexInfo._2.asScala.mkString(","), "enabled")
-          } else {
-            (indexInfo._1, indexInfo._2.asScala.mkString(","), "disabled")
+    val indexesMap = CarbonIndexUtil.getIndexesMap(carbonTable)
+    if (null != indexesMap) {
+      val indexTableMap = indexesMap.asScala
+      if (indexTableMap.nonEmpty) {
+        val secondaryIndex = indexTableMap.get(CarbonIndexProvider.SI.getIndexProviderName)
+        var finalIndexList: Seq[(String, String, String, String, String, String)] = Seq.empty
+
+        if (secondaryIndex.isDefined && null != secondaryIndex.get) {
+          val siIterator = secondaryIndex.get.entrySet().iterator()
+          while (siIterator.hasNext) {
+            val indexInfo = siIterator.next()
+            try {
+              val isSITableEnabled = sparkSession.sessionState.catalog
+                .getTableMetadata(TableIdentifier(indexInfo.getKey, dbNameOp)).storage.properties
+                .getOrElse("isSITableEnabled", "true").equalsIgnoreCase("true")
+              if (isSITableEnabled) {
+                finalIndexList = finalIndexList :+
+                                 (indexInfo.getKey, "carbondata", indexInfo.getValue
+                                   .get(CarbonCommonConstants.INDEX_COLUMNS), "NA", "enabled", "NA")
+              } else {
+                finalIndexList = finalIndexList :+
+                                 (indexInfo.getKey, "carbondata", indexInfo.getValue
+                                   .get(CarbonCommonConstants
+                                     .INDEX_COLUMNS), "NA", "disabled", "NA")
+              }
+            } catch {
+              case ex: Exception =>
+                LOGGER.error(s"Access storage properties from hive failed for index table: ${
+                  indexInfo.getKey
+                }")
+                finalIndexList = finalIndexList :+
+                                 (indexInfo.getKey, "carbondata", indexInfo.getValue
+                                   .get(CarbonCommonConstants.INDEX_COLUMNS), "NA", "UNKNOWN", "NA")
+            }
+          }
+        }
+
+        indexesMap.asScala
+          .filter(map => !map._1.equalsIgnoreCase(CarbonIndexProvider.SI.getIndexProviderName))
```

Comment:
As mentioned above instead of 1 map, if we keep 3 map (each for it's index provider). No need filter logic

Reply:
Instead of 1 Map, if we keep 3 Map, it will occupy more memory right

Reply:
No. Now you have map1(map2(map3)))
SI map - map2(map3)
Cgfgmap - map2(map3)
Instead of 3 level map it is two level map. It saves default map location size of map1.

More over size, no need of lookup and code may be cleaner. You can check.

Reply:
Yes. But if we keep single map with provider and index info, it will be easier to validate index, add index info to hive and drop index. Else, seperately, we need to do for both